ebean-core/src/main/java/io/ebeaninternal/server/persist/DefaultPersister.java

```diff
@@ -642,146 +642,222 @@ public int delete(Class<?> beanType, Object id, Transaction transaction, boolean
-        }
-      }
+  private abstract class DeleteBatchHelp {
+    final BeanDescriptor<?> descriptor;
+    final SpiTransaction transaction;
+    final DeleteMode deleteMode;
+
+    public DeleteBatchHelp(BeanDescriptor<?> descriptor, Transaction transaction, DeleteMode deleteMode) {
+      this.descriptor = descriptor;
+      this.transaction = (SpiTransaction) transaction;
+      this.deleteMode = deleteMode;
     }
 
-    if (t.isPersistCascade()) {
-      // OneToOne exported side with delete cascade
-      BeanPropertyAssocOne<?>[] expOnes = descriptor.propertiesOneExportedDelete();
-      for (BeanPropertyAssocOne<?> expOne : expOnes) {
-        BeanDescriptor<?> targetDesc = expOne.targetDescriptor();
-        // only cascade soft deletes when supported by target
-        if (deleteMode.isHard() || targetDesc.isSoftDelete()) {
-          if (deleteMode.isHard() && targetDesc.isDeleteByStatement()) {
-            SqlUpdate sqlDelete = expOne.deleteByParentId(id, idList);
-            executeSqlUpdate(sqlDelete, t);
-          } else {
-            List<Object> childIds = expOne.findIdsByParentId(id, idList, t);
-            if (childIds != null && !childIds.isEmpty()) {
-              deleteChildrenById(t, targetDesc, childIds, deleteMode);
-            }
-          }
+    int deleteBatch() {
+
+      if (transaction.isPersistCascade()) {
+        BeanPropertyAssocOne<?>[] propImportDelete = descriptor.propertiesOneImportedDelete();
+        if (propImportDelete.length > 0) {
+          return deleteImported(propImportDelete);
         }
-      }
 
-      // OneToMany's with delete cascade
-      BeanPropertyAssocMany<?>[] manys = descriptor.propertiesManyDelete();
-      for (BeanPropertyAssocMany<?> many : manys) {
-        if (!many.isManyToMany()) {
-          BeanDescriptor<?> targetDesc = many.targetDescriptor();
+        // OneToOne exported side with delete cascade
+        BeanPropertyAssocOne<?>[] expOnes = descriptor.propertiesOneExportedDelete();
+        for (BeanPropertyAssocOne<?> expOne : expOnes) {
+          BeanDescriptor<?> targetDesc = expOne.targetDescriptor();
           // only cascade soft deletes when supported by target
           if (deleteMode.isHard() || targetDesc.isSoftDelete()) {
             if (deleteMode.isHard() && targetDesc.isDeleteByStatement()) {
-              // we can just delete children with a single statement
-              SqlUpdate sqlDelete = many.deleteByParentId(id, idList);
-              executeSqlUpdate(sqlDelete, t);
+              executeSqlUpdate(sqlDeleteChildren(expOne), transaction);
             } else {
-              // we need to fetch the Id's to delete (recurse or notify L2 cache)
-              List<Object> childIds = many.findIdsByParentId(id, idList, t, null, deleteMode.isHard());
-              if (!childIds.isEmpty()) {
-                delete(targetDesc, null, childIds, t, deleteMode);
+              List<Object> childIds = findChildIds(expOne, deleteMode.isHard());
```

Comment:
new code: "delteHard" is passed for ones. This could be a functional change, although I think this is correct. Please check

Reply:
I think this is correct too.